### ISSUE TITLE
Elasticsearch6 generation does not exist

### ISSUE THREAD

**Issue by wclshashank** [magento/magento2#21916]
### Preconditions (*)

1. Getting issue while upgrade magento 2.3 to Magento 2.3.1
2. Using Php 7.1.27 version on ubuntu OS

### Steps to reproduce (*)

1. After upgrade magento version getting issue on di:compile like 

> Source class "\Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductField" for "Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper" generation does not exist.

2. I have checked class manually, it doesn't exist.

### Expected result (*)

1. Issue doesn't appear di compile need to work properly.

### Actual result (*)

1. ![image](https://user-images.githubusercontent.com/11975238/54900611-efe38780-4ef9-11e9-8956-3bf3d460fa31.png)
2. Please help us to resolve this issue.

**Comment by magento-engcom-team** [comment]
Hi @wclshashank. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@wclshashank do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @wclshashank the 2.3.1 release will be 26 how can you upgrade ?

**Comment by wclshashank** [comment]
HI @engcom-backlog-nazar ,

Ok got it, I didn't check the release date and try to upgrade magneto2 version. 

Thanks.

**Comment by ghost** [comment]
@wclshashank i'm closing this as magento 2.3.0 do not have Elasticsearch 6 module only develop branch. and on 2.3-develop have no this issue.
![DeepinScreenshot_select-area_20190325094955](https://user-images.githubusercontent.com/40756439/54902991-793b8f80-4ee3-11e9-84f3-439d5aac9484.png)

**Comment by juharintanen** [comment]
Updated to 2.3.1 now and same error still appears.
Only way around it was to disable Magento_Elasticsearch6 module

That file indeed did not exist

**Comment by ghost** [comment]
hi @juharintanen i'm will recheck.

**Comment by wclshashank** [comment]
Hi,

I have this issue due to Amasty module. Amasty module using elasticsearch, which is making issue in compiling.
So i have disable module bin/magento module:disable Amasty_Shopby Amasty_ShopbyPage Amasty_ShopbyBrand
and starts working.

Need to update Amasty module.

Thanks

**Comment by juharintanen** [comment]
> 
> 
> Hi,
> 
> I have this issue due to Amasty module. Amasty module using elasticsearch, which is making issue in compiling.
> So i have disable module bin/magento module:disable Amasty_Shopby Amasty_ShopbyPage Amasty_ShopbyBrand
> and starts working.
> 
> Need to update Amasty module.
> 
> Thanks

Yep confirmed!
It wasn't that but another Amasty module. Error msg just didn't hint that way at all but good to know.

**Comment by ghost** [comment]
@wclshashank  found internal ticket where this was fixed, wait a little bit i'm share with pr.

**Comment by fabio-fafle** [comment]
Hi,
from my understanding if you have a plugin on `Magento\Elasticsearch\Elasticsearch5\Model\Adapter\FieldMapper\ProductFieldMapper` the virtual type `Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper` will inherit it and this cause the issue on method 
`Magento\Setup\Module\Di\Code\Generator\InterceptionConfigurationBuilder::filterNullInheritance`
`if ($plugins === null || !$this->typeReader->isConcrete($instance))`.

It looks like a bug to me.

**Comment by cscd98** [comment]
I have the same issue on 2.3.1:

```
Source class "\Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductField" for "Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper" generation does not exist.
```

Does this require a fix from Magento or from Amasty?

@engcom-backlog-nazar please share the PR

**Comment by bahramdavoodi** [comment]
Hi @craigcarnell,

I have the same error after run di:compile. if you have Extension Amasty Layered Navigation, you can to try to disable extension Amasty_Shopby and run di:compile.

I hope it helps you.
Thanks

**Comment by ghost** [comment]
Hi @craigcarnell why reopen ?

**Comment by cscd98** [comment]
@engcom-backlog-nazar Because a fix has not been provided by Magento

**Comment by cscd98** [comment]
@engcom-backlog-nazar Do you have the PR you mentioned? I need this ASAP please

**Comment by ghost** [comment]
@craigcarnell no, because this in private repository. https://github.com/magento/magento2ce/pull/3912/files#diff-b2e0d454e4ffd542fa96dc1b7dc225d9

**Comment by cscd98** [comment]
@engcom-backlog-nazar Slightly confused, as you offered a fix above. Will create a ticket with Magento "Gold" support..

**Comment by hostep** [comment]
@engcom-backlog-nazar: can you please link to open sourced repo's please? It would be very interesting to see this fix as [other tickets](https://github.com/magento/magento2/issues/21976) are being created which are probably caused by the same issue.

Just link to commits in this repo if they have already been merged, that would be very much appreciated!

**Comment by wclshashank** [comment]
Hi @engcom-backlog-nazar ,

Is there any update?

**Comment by adamrandazzo** [comment]
I can confirm that this seems to be caused by a conflict with the Amasty_ShopBy plugin. I can get it to compile if I disable Amasty_ShopBy OR Magento_Elasticsearch6.

**Comment by ghost** [comment]
@adamrandazzo so to have this issue we need just add a plugin to ElasticSearch 6 ? can you show where plugin defined ?

**Comment by davoc** [comment]
Same issue when upgrading from 2.3.0 to 2.3.1 and same solution here - disable Amasty_Shopby and setup:di:compile completes without error

**Comment by hostep** [comment]
@engcom-backlog-nazar: we would need to figure out what causes this, it's either:
- By introducing ElasticSearch6 module, it caused a backwards incompatibility somehow
- Magento 2.3.1 has a new bug somewhere which causes this
- Amasty's module isn't properly written and contains a bug

I'm guessing it's the second one since you said:
> found internal ticket where this was fixed, wait a little bit i'm share with pr.

But this needs to be verified. Can you maybe give us a summary of what is mentioned in https://github.com/magento/magento2ce/pull/3912/ so we can try to understand if this is a fix for a bug in Magento which can trigger this, or if that PR is something unrelated?

**Comment by ishakhsuvarov** [comment]
@hostep Here is a merge commit for #3912 mentioned above:
https://github.com/magento/magento2/commit/06269b79023d0ba258ae45c4586204f45b5c80dd

**Comment by hostep** [comment]
Thanks @ishakhsuvarov!

So, here's the commit dealing with the elasticsearch6 module problem: https://github.com/magento/magento2/commit/885b7a99397e52e1b094c3d23c314ca508429b3b (which is included in that merge commit), but I'm not sure if this will solve this issue, maybe somebody who has this problem can try adding those lines and see if that fixes anything?

**Comment by hostep** [comment]
I'm pretty sure the commit above will not fix it.

I'm wondering why Magento tries to load the class `Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductField`, which isn't defined anywhere. There is a virtualType `Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper` defined, but that's not the same. It looks like the last part `Mapper` gets stripped of for some reason. This also happens in https://github.com/magento/magento2/issues/21976#issuecomment-477487437

Does anybody have an idea what could trigger this?

**Comment by maghamed** [comment]
hey guys,  @hostep @wclshashank  @adamrandazzo @fabio-fafle 

there is a new release of Amasty Layered Navigation extension available which fixes this issue `v2.11.2` from the extension side. 

https://amasty.com/improved-layered-navigation-for-magento-2.html

Please update the extension and proceed with DI compile. 
Sorry for inconveniences, we will provide a more detailed explanation after the issue investigation soon.

**Comment by hostep** [comment]
Just searching through the diff of Magento 2.3.0 to 2.3.1 looking for the keyword `substr` reveals a bunch of changes in `lib/internal/Magento/Framework/Reflection/TypeProcessor.php`, which seems to be introduced by: https://github.com/magento/magento2/pull/17424/files#diff-292261de777138aa6bb4342b9757a3a5

Could this have anything to do with this (just guessing here)?

**Comment by wclshashank** [comment]
Hi @maghamed ,

Yes, I have upgrade Amasty module now all dependency issue resolved.

Thanks for your help.

**Comment by cscd98** [comment]
Going to close this issue, as Amasty was helpful. Shame about Magento support 👎

**Comment by peterkoppenaal** [comment]
After upgrading the Amasty modules, setup:di:compile works as expected. But when I configure elasticsearch 6 in Magento, indexer:reindex gives me the following error:

`Source class "\Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductField" for "Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper" generation does not exist.`

**Comment by wclshashank** [comment]
Hi @peterkoppenaal ,

Need to check Amasty Module version. It should be 2.11.2. This version working on my project well.
Need to follow:

composer require amasty/shopby ^2.0.0 --update-with-dependencies
and then run composer install

That will resolve your issue.

**Comment by peterkoppenaal** [comment]
@wclshashank I am running on 2.11.2. Composer.lock says:
`{
    "name": "amasty/shopby",
    "version": "2.11.2",
    "dist": {
        "type": "zip",
        "url": "https://composer.amasty.com/enterprise/packages/shopby-2.11.2-ee.zip"
    },
    "require": {
        "amasty/base": ">=1.6.1",
        "amasty/module-improved-layered-navigation-page": ">=2.5.10",
        "amasty/module-shop-by-base": ">=2.10.5",
        "amasty/module-shop-by-brand": ">=2.7.1",
        "amasty/shopby-seo": ">=2.9.12",
        "magento/module-layered-navigation": ">=100.0",
        "php": "~5.6.5|7.0.2|7.0.4|~7.0.6|^7.1.0"
    },
    "type": "magento2-module",
    "autoload": {
        "files": [
            "registration.php"
        ],
        "psr-4": {
            "Amasty\\Shopby\\": ""
        }
    },
    "license": [
        "Commercial"
    ],
    "description": "Improved Layered Navigation by Amasty"
},`

I ran di:compile, cache:clean, cache:flush and indexer:reindex

Error still occurs.

**Comment by wclshashank** [comment]
OK So may be some other module get effected on this. check  di:compile after disable all 3rd party module and custom module.

**Comment by peterkoppenaal** [comment]
@wclshashank This is a new project with only some Amasty modules installed and no Custom work yet.
When I disable Amasty_Shopby it's working.

**Comment by daaannorris** [comment]
Hi all,

I raised this issue with the team at Amasty and they released a patch yesterday evening - after running composer update the issue has been resolved.
Hopefully, this will be the same for you all.

Cheers,
Dan

**Comment by peterkoppenaal** [comment]
@daaannorris Can you run a successful indexer:reindex with Elasticsearch 6 enabled and configured in Magento Admin > Stores > Configuration > Catalog > Catalog Search?

**Comment by daaannorris** [comment]
Hi @peterkoppenaal I don't have Elasticsearch 6 configured so cannot confirm - but I can agree with @wclshashank that upgrading Amasty modules resolves the original issue.
I would suggest raising a ticket with Amasty as your issue appears to relate to their module rather than core Magento - good luck!

**Comment by magento-engcom-team** [comment]
@phoenix128 Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess
- [ ] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by phoenix128** [comment]
I was able to reproduce and it is a general issue related to virtual types with names ending with any autogenerated suffix being pluginized on concrete class.

**Comment by Zyles** [comment]
Same problem. Not using Amasty_ShopBy ...

**Comment by hostep** [comment]
Interesting tidbit: if I copy the Elasticsearch6 module into a Magento 2.3.0 installation with the unpatched Amasty module, the same problem occurs. So this is no new bug in 2.3.1.

I also can confirm what @phoenix128 figured out, in that certain names of virtual types can not be used, if they end on for example `Mapper` or `Logger` or something like that. Then that last part gets stripped of. This happens [over here](https://github.com/magento/magento2/blob/2.3.1/lib/internal/Magento/Framework/Code/Generator.php#L103-L114).
Has this always been the case? Is this expected? Is this documented somewhere?

The [WIP fix](https://github.com/magento/magento2/pull/22046) unfortunately doesn't seem to resolve the problem in my testing.

@maghamed: have you guys already found more info about this problem?

**Comment by akosglue** [comment]
I got the same error on 2.3.1.
`Catalog Search indexer process unknown error:
Source class "\Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductField" for "Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper" generation does not exist.`
After disabling all Amasty extensions the error disappeared.

**Comment by bijalbhavsar** [comment]
This error occurs after upgrading 2.3.1 and it is bug in Magento 2.3.1. If we see https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Elasticsearch6/etc/di.xml it is showing below code:

` <type name="Magento\Elasticsearch\Elasticsearch5\Model\Adapter\FieldMapper\ProductFieldMapperProxy">
        <arguments>
            <argument name="productFieldMappers" xsi:type="array">
                <item name="elasticsearch6" xsi:type="object">Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper</item>
            </argument>
        </arguments>
    </type>`

In above code it states "Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper" as a argument for productFieldMappers and no such model class file exists.  

Review below file path to verify it:  https://github.com/magento/magento2/tree/2.3-develop/app/code/Magento/Elasticsearch6/Model/Adapter/FieldMapper

To fix this issue, either we have to remove this code snippet or create a class file for the same. 

As per me, It is Magento issue and not relevant to Amasty.

**Comment by hostep** [comment]
@bijalbhavsar : `Magento\Elasticsearch6\Model\Adapter\FieldMapper\ProductFieldMapper` exists as a virtual type, see: https://github.com/magento/magento2/blob/2.3.1/app/code/Magento/Elasticsearch6/etc/di.xml#L158

**Comment by phoenix128** [comment]
My PR (https://github.com/magento/magento2/pull/22046/files) is in WIP status, but I am almost there to fix it. Seems like the issue is divided in two different aspects and now I fixed just one of them. More precisely I just fixed it in the developer mode. Need to fix in production mode.

**Comment by phoenix128** [comment]
Update:
I just finalized it. Pull request https://github.com/magento/magento2/pull/22046 seems to fix the issue.

**Comment by magento-engcom-team** [comment]
@sdzhepa Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @sdzhepa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-99115` were created

**Issue Available**: @sdzhepa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @wclshashank. Thank you for your report. 
The issue has been fixed in magento/magento2#22046 by @phoenix128 in 2.3-develop branch
Related commit(s):
 - [fa079c94c9c0a6d3cd7d9dca3b52d3394de50934](https://github.com/magento/magento2/commit/fa079c94c9c0a6d3cd7d9dca3b52d3394de50934)
 - [75684cadfe6231e352c959cbd0648612489b533a](https://github.com/magento/magento2/commit/75684cadfe6231e352c959cbd0648612489b533a)
 - [b8ab61b86f05c45343c78c3341920756f888904f](https://github.com/magento/magento2/commit/b8ab61b86f05c45343c78c3341920756f888904f)
 - [6d873304742ee5114ec661d6b799bd2f0a4f3cff](https://github.com/magento/magento2/commit/6d873304742ee5114ec661d6b799bd2f0a4f3cff)

The fix will be available with the upcoming 2.3.2 release.
